pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java

```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
     if (externalView == null) {
       return;
     }
+
+    // Check if there are relevant changes to the state or if we can skip this update
+    if (!isRoutingTableRebuildRequired(tableName, externalView, instanceConfigList)) {
+      return;
+    }
+
+    // Build the routing table
+    buildRoutingTable(tableName, externalView, instanceConfigList);
+  }
+
+  private boolean isRoutingTableRebuildRequired(String tableName, ExternalView externalView,
+      List<InstanceConfig> instanceConfigs) {
+    // In unit tests, always rebuild the routing table
+    if (_helixManager == null) {
+      return true;
+    }
+
+    // Do we know about this table?
+    if (!_lastKnownExternalViewVersionMap.containsKey(tableName)) {
+      LOGGER.info("Routing table for table {} requires rebuild due to it being newly added", tableName);
+      return true;
+    }
+
+    // Check if the znode version changed
     int externalViewRecordVersion = externalView.getRecord().getVersion();
-    if (_routingTableLastKnownZkVersionMap.containsKey(tableName)) {
-      long lastKnownZkVersion = _routingTableLastKnownZkVersionMap.get(tableName);
-      if (externalViewRecordVersion == lastKnownZkVersion) {
-        LOGGER.info(
-            "No change on routing table version (current version {}, last known version {}), do nothing for table {}",
-            externalViewRecordVersion, lastKnownZkVersion, tableName);
-        return;
+    int lastKnownExternalViewVersion = _lastKnownExternalViewVersionMap.get(tableName);
+
+    if (externalViewRecordVersion != lastKnownExternalViewVersion) {
+      LOGGER.info(
+          "Routing table for table {} requires rebuild due to external view change (current version {}, last known version {})",
+          tableName, externalViewRecordVersion, lastKnownExternalViewVersion);
+      return true;
+    }
+
+    // Check if there are relevant instance config changes
+    Map<String, InstanceConfig> lastKnownInstanceConfigs = _lastKnownInstanceConfigsForTable.get(tableName);
```

Comment:
It just keeps track of the previous instance configs and doesn't rebuild if there are no instance config changes or if the instance config changes are unrelated to `isShutdownInProgress` and `HELIX_ENABLED`. There's unfortunately not a very good compact way to express this in Java.